compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td

```diff
@@ -493,4 +494,59 @@ def IREECodegen_InnerTiledOp : Op<IREECodegen_Dialect, "inner_tiled", [
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// HintWorkgroupCountOp
+//===----------------------------------------------------------------------===//
+
+def IREECodegen_WorkgroupCountHintOp :
+    Op<IREECodegen_Dialect, "workgroup_count_hint", []> {
+  let summary = [{Hints at the workgroup count to set}];
+
+  let description = [{
+    Captures a set of values to use as the workgroup count. The backward slice
+    starting from this op's operands are cloned into the workgroup count region
+    of all transitive callers.
+
+    If `maxDim` is specified, the outer most sizes are linearized down to the
+    `maxDim`th count. `maxDim` is zero-indexed. If multiple hints inform the
+    same entry point, the maximum per dim is used as the count along that dim.
+    For example,
+
+    ```
+    hal.executable.export @entry_point
+    module {
+      func @entry_point() {
+        iree_codegen.workgroup_count_hint sizes(%a, %b) maxDim(0)
+        iree_codegen.workgroup_count_hint sizes(%c, %d, %e) maxDim(1)
+      }
+    }
+    ```
+
+    resolves to
+
+    ```
+    hal.executable.export @entry_point {
+      return max(%a * %b, %c), %d, %e
```

Comment:
I am not immediately sure we want to take multiple hints, and "collapse" them. Also, I think it would be better to just force all the maxDims to be the "least" and then see how to resolve multiple hints. This resolution seems pretty non-intuitive.

Reply:
I think taking individual max this way is ill-defined. I think it is just better off to error out, and only deal with max, when things are linearized.